server/TransactionService.java

```diff
@@ -189,10 +196,13 @@ private void open(TransactionProto.Transaction.Req request) {
         networkLatencyMillis = Math.min(openReq.getNetworkLatencyMillis(), MAX_NETWORK_LATENCY_MILLIS);
         sessionSvc = sessionService(typeDBSvc, openReq);
         sessionSvc.register(this);
-        transaction = transaction(sessionSvc, openReq);
+        options = new Options.Transaction().parent(sessionSvc.options());
+        applyDefaultOptions(options, openReq.getOptions());
+        transaction = transaction(sessionSvc, openReq, options);
```

Comment:
refactored to pass options into the `transaction()` function, so we have access to them a to also set the transaction timeout with

Reply:
This is awkward. Feels weird that you have to pull out the parsing of the `Options` out to a higher level, right? And it feels weird that you have manage the `scheduledTimeout` of the transaction from outside of itself, right?

Given that the transaction is capable of managing its own lifetime, wouldn't it make more sense to get the transaction to manage its own lifetime? It already has access to the timeout limit it should adhere to, and it already has the ability to close itself upon exception. Which means we can encapsulate all the `scheduledTimout` logic inside `RocksTransaction`, right? You would be creating more _cohesion_ without increasing _coupling_, as everything you need to implement the logic is already available in the class. Always take the opportunity to _increase cohesion_ and _reduce coupling_.

Reply:
So I didn't see that before, but I don't think this is very awkward either.
The first part of creating the options first, then passing them into the transaction read fine to me:

1. get session service
2. set up transaction options
3. create the transaction using the session and options

I see your point with the second part, however the intent is actually more that the transaction lifetime is a feature aimed at the RPC, similar to the session idle timeout. This is actually made clearer in the extended version of this feature, outlined in https://github.com/vaticle/typedb/issues/6492. 

Secondly, we currently don't have the architecture for a Transaction (inner component) to pass a sensible error message to its parent (TransactionService) when it times out. We would end up in a (imo) less ideal state that simply logs the transaction timeout error on the server, and throws a "Transaction closed" exception to the client when it next tries to use the transaction.

Reply:
If you think about what RPC means -- Remote Procedure Call, the lifetime of the object really should have nothing to do with the fact that it can be called "remotely".  Which means the session lifetime is probably something that should have been managed by Session itself? I can't be too sure without looking the code directly. However, at this point, I think the ability to send a helpful error message is probably more important. And since we don't want to spend on refactoring our error message architecture, let's move forward with this implementation for now, and refactor it in the next version. But hopefully, the lesson of cohesiveness and coupling can still be taken away.

Reply:
Yep sounds good.
One note on viewing it as an "RPC"  feature, i agree with the definition of RPC you pointed out. Our implementation of "RPC"-based sessions/transactions are referred to as `SessionService` and `TransactionService`, so from those names it wouldn't be strange to have the services manage the lifetimes of the objects they serve either?